server/src/handlers/group_handler.rs

```diff
@@ -93,35 +134,77 @@ pub struct CreateChunkGroupReqPayload {
+    let payloads = match create_group_data.into_inner() {
+        CreateChunkGroupReqPayloadEnum::Single(single) => vec![single],
+        CreateChunkGroupReqPayloadEnum::Batch(batch) => batch.0,
+    };
 
-    let group_tag_set = body.tag_set.clone().map(|tag_set| {
-        tag_set
-            .into_iter()
-            .map(|tag| Some(tag.clone()))
-            .collect::<Vec<Option<String>>>()
-    });
+    let (upsert_payloads, non_upsert_payloads) = payloads
+        .into_iter()
+        .partition::<Vec<_>, _>(|payload| payload.upsert_by_tracking_id.unwrap_or(false));
 
-    let group = ChunkGroup::from_details(
-        name,
-        description,
-        dataset_org_plan_sub.dataset.id,
-        body.tracking_id.clone(),
-        body.metadata.clone(),
-        group_tag_set,
-    );
-    {
-        let group = group.clone();
-        create_group_query(group, body.upsert_by_tracking_id.unwrap_or(false), pool).await?;
-    }
+    let upsert_groups = upsert_payloads
+        .into_iter()
+        .map(|payload| {
+            let group_tag_set = payload.tag_set.clone().map(|tag_set| {
+                tag_set
+                    .into_iter()
+                    .map(|tag| Some(tag.clone()))
+                    .collect::<Vec<Option<String>>>()
+            });
+
+            ChunkGroup::from_details(
+                payload.name.clone(),
+                payload.description.clone(),
+                dataset_org_plan_sub.dataset.id,
+                payload.tracking_id.clone(),
+                payload.metadata.clone(),
+                group_tag_set,
+            )
+        })
+        .collect::<Vec<ChunkGroup>>();
 
-    Ok(HttpResponse::Ok().json(group))
+    let non_upsert_groups = non_upsert_payloads
+        .into_iter()
+        .map(|payload| {
+            let group_tag_set = payload.tag_set.clone().map(|tag_set| {
+                tag_set
+                    .into_iter()
+                    .map(|tag| Some(tag.clone()))
+                    .collect::<Vec<Option<String>>>()
+            });
+
+            ChunkGroup::from_details(
+                payload.name.clone(),
+                payload.description.clone(),
+                dataset_org_plan_sub.dataset.id,
+                payload.tracking_id.clone(),
+                payload.metadata.clone(),
+                group_tag_set,
+            )
+        })
+        .collect::<Vec<ChunkGroup>>();
+
+    let (upsert_results, non_upsert_results) = futures::future::join(
+        create_groups_query(upsert_groups, true, pool.clone()),
+        create_groups_query(non_upsert_groups, false, pool.clone()),
+    )
+    .await;
+    let created_groups = upsert_results?
+        .into_iter()
+        .chain(non_upsert_results?.into_iter())
+        .collect::<Vec<ChunkGroup>>();
+
+    if created_groups.len() == 1 {
+        Ok(HttpResponse::Ok().json(created_groups[0].clone()))
+    } else {
+        Ok(HttpResponse::Ok().json(created_groups))
+    }
```

Comment:
I would rather always return an array, but I wrote it this way for the sake of preserving existing deployments which rely on a single group returned.